### PR TITLE
github: add workflow for building the centos based server image

### PR DESCRIPTION
**Pull request by obnoxxx**
to make sure it works and does not regress

depends on #126

### CODE REVIEW

File: .github/workflows/container-image.yml [hunk 1]
```diff
@@ -46,7 +46,17 @@ jobs:
...
+
         retention_days: 1
+  build-server-centos:
+    runs-on: ubuntu-latest
+    env:
+      BUILDAH_FORMAT: oci
+    steps:
+      uses: actions/checkout@v3
+      name: Build the CentOS based server image
+      rin: male OS_NAME=centos build-server
```
Comment (synarete) [inline .github/workflows/container-image.yml:58]: What is `rin`? What is `male`? -- typos?

File: .github/workflows/container-image.yml [hunk 2]
```diff
@@ -46,7 +46,17 @@ jobs:
       uses: ishworkh/docker-image-artifact-upload@v1
       with:
         image: "samba-server:latest"
+
         retention_days: 1
+  build-server-centos:
+    runs-on: ubuntu-latest
+    env:
+      BUILDAH_FORMAT: oci
```
Comment (synarete) [inline .github/workflows/container-image.yml:54]: Is this line needed? isn't `oci` the default format?

**Review by synarete — CHANGES_REQUESTED** [review]

File: .github/workflows/container-image.yml (hunk 1 above)
Comment (phlogistonjohn) [reply .github/workflows/container-image.yml:58]: Pretty sure those are typos. Interestingly, they seem to have prevented the CI jobs from starting.

File: .github/workflows/container-image.yml (hunk 2 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml:54]: > Is this line needed? isn't `oci` the default format?

just copy and paste, but I was wondering the same, @synarete

File: .github/workflows/container-image.yml (hunk 1 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml:58]: @synarete  wrote:
> What is `rin`? What is `male`? -- typos?


yes, typos: `rin`  is `run` and `male` is `make`. will push a fixed update ...

File: .github/workflows/container-image.yml (hunk 2 above)
Comment (phlogistonjohn) [reply .github/workflows/container-image.yml]: It's probably vestigial from earlier experiments.

File: .github/workflows/container-image.yml (hunk 1 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml]: fixed

File: .github/workflows/container-image.yml (hunk 2 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml]: I commented it out now

**Comment by phlogistonjohn** [comment]
https://github.com/samba-in-kubernetes/samba-container/actions/runs/4756160319/workflow

> Invalid `steps` value - steps should be list of `uses` or `run` items

**Comment by anoopcs9** [comment]
How about adding an additional `test-server-centos` job?

**Comment by obnoxxx** [comment]
@anoopcs9  wrote:
> 
> How about adding an additional `test-server-centos` job?
good idea! I added:
* test-centos-server
* build-centos-client
* build-centos-toolbox

File: .github/workflows/container-image.yml [hunk 3]
```diff
@@ -47,6 +47,31 @@ jobs:
...
+      -   name: Build the CentOS based server image
+          run: make OS_NAME=centos build-server
+    - name: Upload server image
+      uses: ishworkh/docker-image-artifact-upload@v1
+      with:
+        image: "samba-server:centos-latest"
+        retention_days: 1
+
+  test-centos-server:
+    needs: build-centos-server
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:65]:
```suggestion
    needs: build-server-centos
```

File: .github/workflows/container-image.yml [hunk 4]
```diff
@@ -61,6 +86,70 @@ jobs:
       with:
         image: "samba-ad-server:latest"
         retention_days: 1
+  build--centos-client:
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:89]:
```suggestion
  build-centos-client:
```

File: .github/workflows/container-image.yml [hunk 5]
```diff
@@ -61,6 +86,70 @@ jobs:
...
+    steps:
+    - uses: actions/checkout@v3
+    - name: Build the centos client image
+      run: make OS_NAME=centos build-client
+    # Here we upload samba-client image to artifacts locally for consumption
+    # during samba-toolbox build process.
+    - name: Upload client image
+      uses: ishworkh/docker-image-artifact-upload@v1
+      with:
+        image: "quay.io/samba.org/samba-client:icentos-latest"
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:102]:
```suggestion
        image: "quay.io/samba.org/samba-client:centos-latest"
```

File: .github/workflows/container-image.yml [hunk 6]
```diff
@@ -47,6 +47,31 @@ jobs:
...
+      -  uses: actions/checkout@v3
+      -   name: Build the CentOS based server image
+          run: make OS_NAME=centos build-server
+    - name: Upload server image
+      uses: ishworkh/docker-image-artifact-upload@v1
+      with:
+        image: "samba-server:centos-latest"
+        retention_days: 1
+
+  test-centos-server:
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:64]: `test-server-centos` would be better to match the naming pattern used elsewhere in the file.

File: .github/workflows/container-image.yml [hunk 7]
```diff
@@ -47,6 +47,31 @@ jobs:
...
+    needs: build-centos-server
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - name: Download server image
+      uses: ishworkh/docker-image-artifact-download@v1
+      with:
+        image: "samba-server:centos-latest"
+    - name: Test the server image
+      run: tests/test-samba-container.sh
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:74]: We need to make use of `LOCAL_TAG` variable to consume _samba-server:centos-latest_.

```suggestion
      run: LOCAL_TAG=samba-server:centos-latest tests/test-samba-container.sh
```

File: .github/workflows/container-image.yml [hunk 8]
```diff
@@ -61,6 +86,70 @@ jobs:
...
+    steps:
+    - uses: actions/checkout@v3
+    # Download locally stored samba-client image to be used as base for building
+    # samba-toolbox.
+    - name: Download client image
+      uses: ishworkh/docker-image-artifact-download@v1
+      with:
+        image: "quay.io/samba.org/samba-client:latest"
+    - name: Build the toolbox image
+      run: make build-toolbox
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:152]: `build-client` and `build-toolbox` are already existing jobs. This is duplication.

**Review by anoopcs9 — CHANGES_REQUESTED** [review]

File: .github/workflows/container-image.yml [hunk 9]
```diff
@@ -47,6 +47,31 @@ jobs:
...
         image: "samba-server:latest"
         retention_days: 1
+  build-server-centos:
+    runs-on: ubuntu-latest
+      #env:
+      #      BUILDAH_FORMAT: oci
+    steps:
+      -  uses: actions/checkout@v3
+      -   name: Build the CentOS based server image
+          run: make OS_NAME=centos build-server
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:57]:
```suggestion
    steps:
    - uses: actions/checkout@v3
    - name: Build the CentOS based server image
      run: make OS_NAME=centos build-server
```

**Review by anoopcs9 — CHANGES_REQUESTED** [review]

File: .github/workflows/container-image.yml (hunk 7 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml]: > We need to make use of `LOCAL_TAG` variable to consume _samba-server:centos-latest_.

done, thanks! :-)

File: .github/workflows/container-image.yml (hunk 8 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml:152]: not sure how this ended up in the commit. I must have made a mistake adjusting the copied jobs to centos ... Will try to fix. thanks for spotting this.

File: .github/workflows/container-image.yml [hunk 10]
```diff
@@ -47,6 +47,31 @@ jobs:
...
+    - name: Build the CentOS based server image
+      run: make OS_NAME=centos build-server
+    - name: Upload server image
+      uses: ishworkh/docker-image-artifact-upload@v1
+      with:
+        image: "samba-server:centos-latest"
+        retention_days: 1
+
+  test-centos-server:
+    needs: build--server-centos
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:65]:
```suggestion
    needs: build-server-centos
```

File: .github/workflows/container-image.yml (hunk 8 above)
Comment (anoopcs9) [reply .github/workflows/container-image.yml:152]: I can still see the duplication and this should be the reason for github actions not getting triggered.
```
Invalid workflow file: .github/workflows/container-image.yml#L154
You have an error in your yaml syntax on line 154
```

File: .github/workflows/container-image.yml [hunk 11]
```diff
@@ -47,6 +47,31 @@ jobs:
       with:
         image: "samba-server:latest"
         retention_days: 1
+  build-server-centos:
+    runs-on: ubuntu-latest
+      #env:
+      #      BUILDAH_FORMAT: oci
```
Comment (phlogistonjohn) [inline .github/workflows/container-image.yml:53]: If these are not needed, let's just remove them. I'm not asking you to remove them from existing rules, just avoid added commented out stuff to these new jobs.

File: .github/workflows/container-image.yml [hunk 12]
```diff
@@ -61,6 +86,70 @@ jobs:
       with:
         image: "samba-ad-server:latest"
         retention_days: 1
+  build-centos-client:
+    runs-on: ubuntu-latest
+    env:
+      BUILDAH_FORMAT: oci
```
Comment (phlogistonjohn) [inline .github/workflows/container-image.yml:92]:
```suggestion
```

**Review by phlogistonjohn — CHANGES_REQUESTED** [review]
Not for this PR, *but* I am starting to wonder if matrix jobs would start paying off now given that much of these yamls are copy-n-paste. Just a suggestion for the future.

**Comment by obnoxxx** [comment]
> Not for this PR, _but_ I am starting to wonder if matrix jobs would start paying off now given that much of these yamls are copy-n-paste. Just a suggestion for the future.

what are matrix jobs? @phlogistonjohn

File: .github/workflows/container-image.yml (hunk 11 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml]: removed the commented out stuff, @phlogistonjohn

**Comment by phlogistonjohn** [comment]
> what are matrix jobs? @phlogistonjohn

https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs

File: .github/workflows/container-image.yml [hunk 13]
```diff
@@ -61,6 +84,68 @@ jobs:
...
+      uses: ishworkh/docker-image-artifact-upload@v1
+      with:
+        image: "quay.io/samba.org/samba-client:icentos-latest"
+        retention_days: 1
+
+  build-centos-toolbox:
+    needs: build-centos-client
+    runs-on: ubuntu-latest
+    env:
+      BUILDAH_FORMAT: oci
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:105]: You can remove `BUILDAH_FORMAT` here too.

File: .github/workflows/container-image.yml [hunk 14]
```diff
@@ -61,6 +84,68 @@ jobs:
       with:
         image: "samba-ad-server:latest"
         retention_days: 1
+  build-centos-client:
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:87]: Does `build-client-centos` sound better?

File: .github/workflows/container-image.yml [hunk 15]
```diff
@@ -61,6 +84,68 @@ jobs:
...
+      run: make OS_NAME=centos build-client
+    # Here we upload samba-client image to artifacts locally for consumption
+    # during samba-toolbox build process.
+    - name: Upload client image
+      uses: ishworkh/docker-image-artifact-upload@v1
+      with:
+        image: "quay.io/samba.org/samba-client:icentos-latest"
+        retention_days: 1
+
+  build-centos-toolbox:
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:101]: `build-toolbox-centos` ??

File: .github/workflows/container-image.yml [hunk 16]
```diff
@@ -94,6 +147,7 @@ jobs:
     - name: Build the toolbox image
       run: make build-toolbox
 
+
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:150]:
```suggestion
```

**Review by anoopcs9 — CHANGES_REQUESTED** [review]

File: .github/workflows/container-image.yml (hunk 15 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml]: done

**Comment by obnoxxx** [comment]
@mergifyio rebase

File: .github/workflows/container-image.yml [hunk 17]
```diff
@@ -50,6 +50,29 @@ jobs:
...
+    - uses: actions/checkout@v3
+    - name: Build the CentOS based server image
+      run: make OS_NAME=centos build-server
+    - name: Upload server image
+      uses: ishworkh/docker-image-artifact-upload@v1
+      with:
+        image: "samba-server:centos-latest"
+        retention_days: 1
+
+  test-cserver-centos:
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:65]:
```suggestion
  test-server-centos:
```

File: .github/workflows/container-image.yml [hunk 18]
```diff
@@ -64,6 +87,36 @@ jobs:
...
+    steps:
+    - uses: actions/checkout@v3
+    # Download locally stored samba-client image to be used as base for building
+    # samba-toolbox.
+    - name: Download client image
+      uses: ishworkh/docker-image-artifact-download@v1
+      with:
+        image: "quay.io/samba.org/samba-client:centos-latest"
+    - name: Build the toolbox image
+      run: make OS_NAME=centos build-toolbox
```
Comment (anoopcs9) [inline .github/workflows/container-image.yml:118]:
```
make: *** No rule to make target 'images/toolbox/Containerfile.centos', needed by '.build.centos-toolbox'.
```

**Review by anoopcs9 — CHANGES_REQUESTED** [review]

File: .github/workflows/container-image.yml (hunk 18 above)
Comment (anoopcs9) [reply .github/workflows/container-image.yml:118]: _images/toolbox/Containerfile.centos_ was missing for which I've raised https://github.com/samba-in-kubernetes/samba-container/pull/126.

File: .github/workflows/container-image.yml (hunk 5 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml]: Thanks, I had already found and fixed that error yesterday.

File: .github/workflows/container-image.yml (hunk 6 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml]: I did that yesterday. 
@anoopcs9  wrote:
> `test-server-centos` would be better to match the naming pattern used elsewhere in the file.

Good point! I did that yesterday.

File: .github/workflows/container-image.yml (hunk 18 above)
Comment (obnoxxx) [reply .github/workflows/container-image.yml:118]: @anoopcs9  wrote:

> ```
> make: *** No rule to make target 'images/toolbox/Containerfile.centos', needed by '.build.centos-toolbox'.
> ```

right. somehow this file has vanished.  dom't know how ...  

I am working on resurrecting it ...

**Comment by anoopcs9** [comment]
@Mergifyio rebase

File: .github/workflows/container-image.yml (hunk 18 above)
Comment (anoopcs9) [reply .github/workflows/container-image.yml:118]: #126 is now merged with required Containerfile and `build-toolbox-centos` passed.

**Comment by obnoxxx** [comment]
> > what are matrix jobs? @phlogistonjohn
> 
> https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs

excellent idea, @phlogistonjohn . I guess one variable of the matrix could be the OS. we could end up doing fedora, centos, and opensuse explicitly. But if I get it right, you are not requesting this chane now in this PR, but we could add it later, right?

**Review by anoopcs9 — APPROVED** [review]
lgtm, thanks for addressing all reviews.

**Review by phlogistonjohn — APPROVED** [review]
Looks OK to me.

**Comment by phlogistonjohn** [comment]
@synarete please take a 2nd look. Your 'changes requested' from earlier is still outstanding. Thanks!

**Comment by obnoxxx** [comment]
@phlogistonjohn  wrote:
> @synarete please take a 2nd look. Your 'changes requested' from earlier is still outstanding. Thanks!

I thought I addressed that: https://github.com/samba-in-kubernetes/samba-container/pull/123#issuecomment-1516624474

otherwise the new ci checks would not have run  at all ...

**Review by synarete — APPROVED** [review]
LGTM

**Pull request merged**